server/router/src/main/java/org/infinispan/server/router/router/impl/hotrod/HotRodRouter.java

```diff
@@ -0,0 +1,101 @@
+package org.infinispan.server.router.router.impl.hotrod;
+
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.epoll.EpollServerSocketChannel;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.ServerSocketChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.util.concurrent.DefaultThreadFactory;
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.server.router.RoutingTable;
+import org.infinispan.server.router.configuration.HotRodRouterConfiguration;
+import org.infinispan.server.router.logging.RouterLogger;
+import org.infinispan.server.router.router.Router;
+import org.infinispan.server.router.router.impl.hotrod.handlers.SniHandlerInitializer;
+
+import java.lang.invoke.MethodHandles;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.Optional;
+
+public class HotRodRouter implements Router {
+
+    private static final RouterLogger logger = LogFactory.getLog(MethodHandles.lookup().lookupClass(), RouterLogger.class);
+
+    private static final String THREAD_NAME_PREFIX = "MultiTenantRouter";
+
+    private final NioEventLoopGroup masterGroup = new NioEventLoopGroup(1, new DefaultThreadFactory(THREAD_NAME_PREFIX + "ServerMaster"));
+    private final NioEventLoopGroup workerGroup = new NioEventLoopGroup(0, new DefaultThreadFactory(THREAD_NAME_PREFIX + "ServerWorker"));
+    private final HotRodRouterConfiguration configuration;
+    private Optional<Integer> port = Optional.empty();
+    private Optional<InetAddress> ip = Optional.empty();
+
+    public HotRodRouter(HotRodRouterConfiguration configuration) {
+        this.configuration = configuration;
+    }
+
+    @Override
+    public void start(RoutingTable routingTable) {
+        try {
+            ServerBootstrap bootstrap = new ServerBootstrap();
+            bootstrap.group(masterGroup, workerGroup)
+                    .childOption(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
+                    .childOption(ChannelOption.TCP_NODELAY, configuration.tcpNoDelay())
+                    .childOption(ChannelOption.SO_KEEPALIVE, configuration.keepAlive())
+                    .childHandler(new SniHandlerInitializer(routingTable))
+                    .channel(NioServerSocketChannel.class);
+            if (configuration.sendBufferSize() > 0)
+                bootstrap.childOption(ChannelOption.SO_SNDBUF, configuration.sendBufferSize());
+            if (configuration.receiveBufferSize() > 0)
+                bootstrap.childOption(ChannelOption.SO_RCVBUF, configuration.receiveBufferSize());
+
+            InetAddress ip = configuration.getIp();
+            int port = configuration.getPort();
+
+            Channel channel = bootstrap.bind(ip, port).sync().channel();
```

Comment:
I'm not 100% sure to be honest...

Looking at the [user guide](http://netty.io/wiki/user-guide-for-4.x.html) there is no need to shutdown the channel. Here is a quote from a [paragraph about shutdown](http://netty.io/wiki/user-guide-for-4.x.html#shutting-down-your-application):

> Shutting down a Netty application is usually as simple as shutting down all EventLoopGroups you created via shutdownGracefully(). It returns a Future that notifies you when the EventLoopGroup has been terminated completely and all Channels that belong to the group have been closed.

Based on that I think the implementation is fine but I could be easily convinced that I've been wrong ;)